langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/FromJsonWithType.java

```diff
@@ -0,0 +1,268 @@
+
+/**
+ * Extern function lang.values:fromJsonWithType.
+ *
+ * @since 2.0
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "lang.value", version = VALUE_VERSION,
+        functionName = "fromJsonWithType",
+        isPublic = true
+)
+public class FromJsonWithType {
+    private static final String AMBIGUOUS_TARGET = "ambiguous target type";
+
+    public static Object fromJsonWithType(Strand strand, Object v, TypedescValue t) {
+        BType describingType = t.getDescribingType();
+        try {
+            return convert(v, describingType, new ArrayList<>(), t, strand);
+        } catch (ErrorValue e) {
+            return e;
+        } catch (BallerinaException e) {
+            return createError(VALUE_LANG_LIB_CONVERSION_ERROR, e.getDetail());
+        }
+    }
+
+    private static Object convert(Object value, BType targetType, List<TypeValuePair> unresolvedValues,
+                                  TypedescValue t, Strand strand) {
+
+        TypeValuePair typeValuePair = new TypeValuePair(value, targetType);
+        BType sourceType = TypeChecker.getType(value);
+
+        if (unresolvedValues.contains(typeValuePair)) {
+            throw new BallerinaException(VALUE_LANG_LIB_CYCLIC_VALUE_REFERENCE_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CYCLIC_VALUE_REFERENCE, sourceType));
+        }
+
+        unresolvedValues.add(typeValuePair);
+
+        if (value == null) {
+            if (targetType.isNilable()) {
+                return null;
+            }
+            throw createError(VALUE_LANG_LIB_CONVERSION_ERROR,
+                    BLangExceptionHelper.getErrorMessage(RuntimeErrors.CANNOT_CONVERT_NIL, targetType));
+        }
+
+        List<BType> convertibleTypes = TypeConverter.getConvertibleTypesFromJson(value, targetType, new ArrayList<>());
+        if (convertibleTypes.isEmpty()) {
+            throw createConversionError(value, targetType);
+        } else if (convertibleTypes.size() > 1) {
+            throw createConversionError(value, targetType, AMBIGUOUS_TARGET);
+        }
+
+        BType matchingType = convertibleTypes.get(0);
+
+        Object newValue;
+        switch (sourceType.getTag()) {
+            case TypeTags.MAP_TAG:
+            case TypeTags.RECORD_TYPE_TAG:
+                newValue = convertMap((MapValue<?, ?>) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.TUPLE_TAG:
+                newValue = convertArray((ArrayValue) value, matchingType, unresolvedValues, t, strand);
+                break;
+            case TypeTags.ERROR_TAG:
+                newValue = ((RefValue) value).copy(new HashMap<>());
+                break;
+            default:
+                if (((matchingType.getTag() >= TypeTags.XML_ELEMENT_TAG &&
+                        matchingType.getTag() <= TypeTags.XML_TEXT_TAG) || matchingType.getTag() == TypeTags.XML_TAG)) {
```

Comment:
Can't we use `org.ballerinalang.jvm.types.TypeTags#isXMLTypeTag` here instead?

Reply:
Sure